numba/core/types/function_type.py

```diff
@@ -0,0 +1,200 @@
+    def name(self):
+        return f'{type(self).__name__}[{self.key}]'
+
+    def is_precise(self):
+        return self.signature.is_precise()
+
+    def dump(self, tab=''):
+        print(f'{tab}DUMP {type(self).__name__}[code={self._code}]')
+        self.signature.dump(tab=tab + '  ')
+        print(f'{tab}END DUMP {type(self).__name__}')
+
+    def get_call_type(self, context, args, kws):
+        from numba.core import typing
+
+        if len(args) != self.nargs:
+            raise ValueError(
+                f'mismatch of arguments number: {len(args)} vs {self.nargs}')
+
+        sig = self.signature
+
+        # check that arguments types match with the signature types exactly
+        for atype, sig_atype in zip(args, sig.args):
+            atype = types.unliteral(atype)
+            if sig_atype.is_precise():
+                conv_score = context.context.can_convert(
+                    fromty=atype, toty=sig_atype
+                )
+                if conv_score is None \
+                   or conv_score > typing.context.Conversion.safe:
+                    raise ValueError(
+                        f'mismatch of argument types: {atype} vs {sig_atype}')
+
+        if not sig.is_precise():
+            for dispatcher in self.dispatchers:
+                template, pysig, args, kws \
+                    = dispatcher.get_call_template(args, kws)
+                new_sig = template(context.context).apply(args, kws)
+                return types.unliteral(new_sig)
+
+        return sig
+
+    def check_signature(self, other_sig):
+        """Return True if signatures match (up to being precise).
+        """
+        sig = self.signature
+        return (self.nargs == len(other_sig.args)
+                and (sig == other_sig or not sig.is_precise()))
+
+    def unify(self, context, other):
+        if isinstance(other, types.UndefinedFunctionType) \
+           and self.nargs == other.nargs:
+            return self
+
+
+class UndefinedFunctionType(FunctionType):
+
+    _counter = 0
+
+    def __init__(self, nargs, dispatchers):
+        from numba.core.typing.templates import Signature
+        signature = Signature(types.undefined,
+                              (types.undefined,) * nargs, recvr=None)
+
+        super(UndefinedFunctionType, self).__init__(signature)
+
+        self.dispatchers = dispatchers
+
+        type(self)._counter += 1
+        self._key_count = str(type(self)._counter)
+
+    @property
+    def key(self):
+        return self.ftype.key + self._key_count
+
+    def get_precise(self):
+        """
+        Return precise function type if possible.
+        """
+        sig = None
+        for dispatcher in self.dispatchers:
+            for cres in dispatcher.overloads.values():
+                sig = types.unliteral(cres.signature)
+                break
+            else:
+                continue
+            break
+        if sig is None:
+            return self
+
+        self._reset_signature(sig)
```

Comment:
A type object should not mutate itself such that it's `.key` changes.  `.key` is used to compute the hash of the type.  A non-precise type should never become a precise type.  They should define a `.refine()` method that creates a new type that is the precise version.

Reply:
Fixed.
Btw, when reading the `refine` usage, I noticed:
https://github.com/numba/numba/blob/5364a294ce47dfc97628d3fe82b77441d0dab8a5/numba/core/types/containers.py#L524
that does nothing. Is this a bug and be `assert res.is_precise()` or similar?
Another case is
https://github.com/numba/numba/blob/5364a294ce47dfc97628d3fe82b77441d0dab8a5/numba/core/types/containers.py#L606

Reply:
right, there should be an `assert`

Reply:
Fixed by turning the is_precise calls into assert statements.